src/kernels/MIOpenSparseSoftmaxCrossEntropyWithLogits.cpp

```diff
@@ -0,0 +1,280 @@
+    {
+        FLOAT_ACCUM val = CVT_FLOAT2ACCUM(input[input_tv.get_tensor_view_idx({gid, i})]);
+        FLOAT_ACCUM backprop_val =
+            (i == label) ? exp(val - lmax[0]) / lsum[0] - 1.0f : exp(val - lmax[0]) / lsum[0];
+
+        backprop[backprop_tv.get_tensor_view_idx({gid, i})] = CVT_ACCUM2FLOAT(backprop_val);
+    }
+}
+
+extern "C" __global__ void SparseSoftmaxCrossEntropyWithLogitsForward(const D_TYPE* input,
+                                                                      const T_TYPE* target,
+                                                                      D_TYPE* output,
+                                                                      D_TYPE* backprop,
+                                                                      uint64_t num_class,
+                                                                      tensor_view_t<2> input_tv,
+                                                                      tensor_view_t<1> target_tv,
+                                                                      tensor_view_t<1> output_tv,
+                                                                      tensor_view_t<2> backprop_tv)
+{
+    sparseSoftmaxCrossEntropyWithLogitsForward<D_TYPE, T_TYPE>(
+        input, target, output, backprop, num_class, input_tv, target_tv, output_tv, backprop_tv);
+}
+
+template <typename T, typename Ta>
+__device__ void sparseSoftmaxCrossEntropyWithLogitsForwardContiguous(
+    const T* input, const Ta* target, T* output, T* backprop, uint64_t num_class)
+{
+    uint64_t gid = blockIdx.x;
+    uint64_t lid = threadIdx.x;
+
+    __shared__ FLOAT_ACCUM lmax[LOCAL_SIZE], lsum[LOCAL_SIZE];
+    lmax[lid] = DTYPE_LOWEST;
+    lsum[lid] = 0.0f;
+    __shared__ uint64_t label;
+    uint64_t batch_offset = gid * num_class;
+
+    if(lid == 0)
+    {
+        label = static_cast<uint64_t>(target[gid]);
+    }
+
+    for(uint64_t i = lid; i < num_class; i += LOCAL_SIZE)
+    {
+        FLOAT_ACCUM val = CVT_FLOAT2ACCUM(input[i + batch_offset]);
+        lmax[lid]       = max(lmax[lid], val);
+    }
+    __syncthreads();
+
+    for(uint64_t i = LOCAL_SIZE >> 1; i > 0; i >>= 1)
+    {
+        if(lid < i)
+        {
+            lmax[lid] = max(lmax[lid], lmax[lid + i]);
+        }
+        __syncthreads();
+    }
+
+    for(uint64_t i = lid; i < num_class; i += LOCAL_SIZE)
+    {
+        FLOAT_ACCUM val = CVT_FLOAT2ACCUM(input[i + batch_offset]);
+        lsum[lid] += exp(val - lmax[0]);
+    }
+    __syncthreads();
+
+    for(uint64_t i = LOCAL_SIZE >> 1; i > 0; i >>= 1)
+    {
+        if(lid < i)
+        {
+            lsum[lid] += lsum[lid + i];
+        }
+        __syncthreads();
+    }
+
+    for(uint64_t i = lid; i < num_class; i += LOCAL_SIZE)
+    {
+        if(i == label)
+        {
+            FLOAT_ACCUM val = CVT_FLOAT2ACCUM(input[i + batch_offset]);
+            output[gid]     = CVT_ACCUM2FLOAT(log(lsum[0]) - val + lmax[0]);
+        }
+    }
```

Comment:
It's obiviously that their is one label value only. Is the for loop needed?

Reply:
output_grad_val